python/tskit/trees.py

```diff
@@ -4820,10 +4826,25 @@ def samples(self, population=None, population_id=None):
         if population_id is not None:
             population = population_id
         samples = self._ll_tree_sequence.get_samples()
+        keep = np.full(shape=samples.shape, fill_value=True)
         if population is not None:
             sample_population = self.tables.nodes.population[samples]
-            samples = samples[sample_population == population]
-        return samples
+            keep = np.logical_and(keep, sample_population == population)
+        if time is not None:
+            if isinstance(time, (int, float)):
+                sample_times = self.tables.nodes.time[samples]
+                keep = np.logical_and(keep, np.isclose(sample_times, time))
+            elif isinstance(time, (tuple, list)) and len(time) == 2:
```

Comment:
I think it'd be better to coerce the input to a numpy array, something like
```python
else:
     time = np.ndarray(time, dtype=float) # This should raise TypeErrors/ValueErrors as needed
     if time.shape != (2,):
         raise ValueError("time interval must be a pair of values (min_time, max_time)")
     # as before
```

We want to make sure that people can pass in numpy arrays as parameters, which the current approach doesn't allow.

An alternative is to do something like
```python
else:
     min_time, max_time = time 
     if not (isinstance(min_time, numbers.Number) and isinstance(max_time, numbers.Number)):
          raise TypeError(...)
```
which is perhaps clearer and more in line with future changes in Python via [PEP 622](https://www.python.org/dev/peps/pep-0622/)

Any thoughts here anyone? @benjeffery?

Reply:
I agree it is better to coerce the input into a numpy array. So now I'm doing that from the get-go (note I'm using `np.array` and not `np.ndarray` because the latter is not used for array creation).

I opted to check for the shape explicitly as (1,) and (2,) so that we can give a more informative error in the case the input shape is wonky.